datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java

```diff
@@ -151,10 +155,50 @@ public class ZkAdapter {
   private ZkBackedDMSDatastreamList _datastreamList = null;
   private ZkTargetAssignmentProvider _targetAssignmentProvider = null;
   private ZkBackedLiveInstanceListProvider _liveInstancesProvider = null;
+  private ZKThroughputViolatingTopicsProvider _throughputViolatingTopicsProvider = null;
 
   // Cache all live DatastreamTasks per instance for assignment strategy
   private Map<String, Set<DatastreamTask>> _liveTaskMap = new HashMap<>();
 
+  // Cache all the throughput violating topics per datastream
+  private final Map<String, Set<String>> _throughputViolatingTopicsMap = new HashMap<>();
+
+  // As the _throughputViolatingTopicsMap may be updated by watcher thread and read by multiple task threads concurrently.
+  private final ReadWriteLock _throughputViolatingTopicsMapReadWriteLock = new ReentrantReadWriteLock();
+  private final Lock _throughputViolatingTopicsMapWriteLock = _throughputViolatingTopicsMapReadWriteLock.writeLock();
+  private final Lock _throughputViolatingTopicsMapReadLock = _throughputViolatingTopicsMapReadWriteLock.readLock();
+
+
+  /**
+   * ThroughputViolations class keeps track of the throughput violations per datastream and
+   * also provides a wrapper for JSON SerDe.
```

Comment:
Can you get this class verified by Jhora or Eric? Eric suggested a cleaner way in recent PR and  lets check that we are doing that here as well.